src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -552,9 +572,22 @@ public void Flush()
             Output.Flush();
         }
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
+        public Task FlushAsync(CancellationToken cancellationToken)
         {
-            await InitializeResponse(0);
+            var initializeTask = InitializeResponseMayAwait(0);
+            // If return is TaskCache.CompletedTask no awaiting is required
+            if (!ReferenceEquals(initializeTask, TaskCache.CompletedTask))
```

Comment:
[`Task.Status`](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L1343-L1365) never inlines (https://github.com/dotnet/corefx/issues/16745#issuecomment-292728779) because its a bit chunky so going for something simpler to get max gains to make more worthwhile as splitting the functions is quite messy.

The `TaskCache.CompletedTask` field is always used for all the sync completing methods in Kestrel so is an easy equality check. Downside is it won't pick up on an external event (like in onStarted) that completes sync whereas `RanToCompletion` would.